### PR TITLE
WT-7192 Fixes for cached cursor sweep.

### PR DESCRIPTION
**Pull request by ddanderson**
Simplify the WT_DHANDLE_CAN_REOPEN macro, it will no longer return true for some
weird flag combinations.  Use this macro when reopening as a final check.
When doing the sweep check, even if the cursor is otherwise
"reopenable", don't allow sweeping of a cursor that is using the current
dhandle. Remove the assertion that used to protect that case, as it
can no longer happen.

### CODE REVIEW

**Comment by ddanderson** [comment]
@keitharnoldsmith and @sueloverso, I've marked you both as reviewers on this, as there's some subtle stuff at play.  The original analysis I did for the problem is in: https://jira.mongodb.org/browse/WT-7143. (See my long comment).  The changes in this PR are to do Remedy 1 & 2 from my comment, and (I hope) some other small simplifications and comment clarifications.  And then the fix -- recognizing there's a special case when the cursor we're looking at sweeping is the current dhandle.  Let me know if there's anything about this you don't get, we can get into more detail.  Thanks!

File: src/cursor/cur_file.c [hunk 1]
```diff
@@ -608,30 +608,26 @@ __curfile_reopen_int(WT_CURSOR *cursor)
...
-        can_reopen = WT_DHANDLE_CAN_REOPEN(dhandle);
+    if (sweep_check_only) {
         /*
-         * If the data handle can't be reopened, it is a candidate for sweep, and that should never
-         * happen if any session (including ours) is actively using it.
+         * The sweep check returns WT_NOTFOUND if the cursor should be swept. Generally if the
+         * associated data handle cannot be reopened we want to sweep it. But we also don't want to
+         * sweep a handle that we're currently operating on. That can occur if we're in the middle
+         * of a close on a cursor for a handle that is marked as dropped, as session cursor sweeps
+         * can happen while closing.
```
Comment (sueloverso) [inline src/cursor/cur_file.c:627]: I would rephrase some of this comment because I'm getting confused about who "we" means. Here's my attempt at understanding in explicit words. It may not be quite what you want and it may not be accurate. Please fix accordingly.

The sweep check returns WT_NOTFOUND if the cursor should be swept. Generally if the
associated data handle cannot be reopened it should be swept. But a handle being operated
on by this thread should not be swept. The situation where a handle cannot be reopened but
also cannot be swept can occur if this thread is in the middle
of closing a cursor for a handle that is marked as dropped at the same time the connection sweep
thread runs and tries to sweep session cursors.

**Review by sueloverso — APPROVED** [review]
LGTM. This feels a lot cleaner. Some comments on the comment itself.

File: src/cursor/cur_file.c (hunk 1 above)
Comment (ddanderson) [reply src/cursor/cur_file.c:627]: I'm using your wording, except for the very end.  The session sweep is actually running here.  The call stack is roughly:
```
    SESSION->close()  calls...
    release cursor.      calls....
    run a few iterations of the session sweep
    call cursor->reopen (__curfile_reopen) to see if we can sweep this cursor
```

Comment (ddanderson) [reply src/cursor/cur_file.c:627]: Alternate wording:
```
        /*
         * The sweep check returns WT_NOTFOUND if the cursor should be swept. Generally if the
         * associated data handle cannot be reopened it should be swept. But a handle being operated
         * on by this thread should not be swept. The situation where a handle cannot be reopened
         * but also cannot be swept can occur if this thread is in the middle of closing a cursor
         * for a handle that is marked as dropped. During the close, a few iterations of
         * the session cursor sweep are run.  The sweep calls this function to see if a cursor
         * should be swept, and it may thus be asking about the very cursor being closed.
         */
```

**Review by sueloverso — APPROVED** [review]
LGTM. That is much better and I understand it now!

**Review by keitharnoldsmith — APPROVED** [review]
LGTM

I realize it would be pretty tricky to pull off, but is there testing we can/should add to make sure this continues to behave correctly in the future?

**Comment by ddanderson** [comment]
@keitharnoldsmith 
> I realize it would be pretty tricky to pull off, but is there testing we can/should add to make sure this continues to behave correctly in the future?

Interesting question.  I took a quick stab at creating a python test to trigger the failure, without any luck. Since I really want to resolve the steady stream of BFs for this, I'm going to merge and create a ticket to pursue this idea.

**Comment by ddanderson** [comment]
@keitharnoldsmith, I created https://jira.mongodb.org/browse/WT-7194 to track the question you raised.  Thanks!
